File: falcon_mcp/modules/hosts.py
"""
Hosts module for Falcon MCP Server

This module provides tools for accessing and managing CrowdStrike Falcon hosts/devices.
"""

from textwrap import dedent
from typing import Any, Literal

from mcp.server import FastMCP
from mcp.server.fastmcp.resources import TextResource
from pydantic import AnyUrl, Field
from pydantic.fields import FieldInfo

from falcon_mcp.common.field_presets import HOST_SUMMARY_FIELDS
from falcon_mcp.common.logging import get_logger
from falcon_mcp.common.utils import filter_records, format_response
from falcon_mcp.modules.base import BaseModule
from falcon_mcp.resources.hosts import SEARCH_HOSTS_FQL_DOCUMENTATION

logger = get_logger(__name__)


class HostsModule(BaseModule):
    """Module for accessing and managing CrowdStrike Falcon hosts/devices."""

    def register_tools(self, server: FastMCP) -> None:
        """Register tools with the MCP server.

        Args:
            server: MCP server instance
        """
        # Register tools
        self._add_tool(
            server=server,
            method=self.search_hosts,
            name="search_hosts",
        )

        self._add_tool(
            server=server,
            method=self.get_host_details,
            name="get_host_details",
        )

    def register_resources(self, server: FastMCP) -> None:
        """Register resources with the MCP server.

        Args:
            server: MCP server instance
        """
        search_hosts_fql_resource = TextResource(
            uri=AnyUrl("falcon://hosts/search/fql-guide"),
            name="falcon_search_hosts_fql_guide",
            description="Contains the guide for the `filter` param of the `falcon_search_hosts` tool.",
            text=SEARCH_HOSTS_FQL_DOCUMENTATION,
        )

        self._add_resource(
            server,
            search_hosts_fql_resource,
        )

    def search_hosts(
        self,
        filter: str | None = Field(
            default=None,
            description="FQL filter expression. See `falcon://hosts/search/fql-guide` for syntax.",
            examples={"platform_name:'Windows'", "hostname:'PC*'"},
        ),
        limit: int = Field(
            default=10,
            ge=1,
            le=5000,
            description="The maximum records to return. [1-5000]",
        ),
        offset: int | None = Field(
            default=None,
            description="The offset to start retrieving records from.",
        ),
        sort: str | None = Field(
            default=None,
            description=dedent("""
                Sort hosts using these options:

                hostname: Host name/computer name
                last_seen: Timestamp when the host was last seen
                first_seen: Timestamp when the host was first seen
                modified_timestamp: When the host record was last modified
                platform_name: Operating system platform
                agent_version: CrowdStrike agent version
                os_version: Operating system version
                external_ip: External IP address

                Sort either asc (ascending) or desc (descending).
                Both formats are supported: 'hostname.desc' or 'hostname|desc'

                Examples: 'hostname.asc', 'last_seen.desc', 'platform_name.asc'
            """).strip(),
            examples={"hostname.asc", "last_seen.desc"},
        ),
        view: Literal["summary", "full"] = Field(
            default="summary",
            description="'summary' returns investigation-essential fields only (default). 'full' returns the complete API response.",
        ),
        fields: list[str] | None = Field(
            default=None,
            description="Override: explicit list of fields to return. Takes precedence over view.",
        ),
        format: Literal["json", "toon"] = Field(
            default="json",
            description="Response format. 'toon' uses compact tabular encoding for token efficiency.",
        ),
    ) -> list[dict[str, Any]] | str:
        """Search for hosts in your CrowdStrike environment.

        Use this to find devices by hostname, platform, IP, sensor version, or other
        attributes. Consult falcon://hosts/search/fql-guide before constructing filter
        expressions. Returns full host details including device info, OS, and network
        context.
        """
        # Resolve FieldInfo defaults for direct calls (e.g., from tests)
        if isinstance(view, FieldInfo):
            view = view.default
        if isinstance(fields, FieldInfo):
            fields = fields.default
        if isinstance(format, FieldInfo):
            format = format.default

        device_ids = self._base_search_api_call(
            operation="QueryDevicesByFilter",
            search_params={
                "filter": filter,
                "limit": limit,
                "offset": offset,
                "sort": sort,
            },
            error_message="Failed to search hosts",
        )

        # If handle_api_response returns an error dict instead of a list,
        # it means there was an error, so we return it wrapped in a list
        if self._is_error(device_ids):
            return [device_ids]

        # If we have device IDs, get the details for each one
        if device_ids:
            details = self._base_get_by_ids(
                operation="PostDeviceDetailsV2",
                ids=device_ids,
                id_key="ids",
            )

            # If handle_api_response returns an error dict instead of a list,
            # it means there was an error, so we return it wrapped in a list
            if self._is_error(details):
                return [details]

            if fields:
                details = filter_records(details, fields)
            elif view == "summary":
                details = filter_records(details, HOST_SUMMARY_FIELDS)

            return format_response(details, format)

        return []

    def get_host_details(
        self,
        ids: list[str] = Field(
            description="Host device IDs to retrieve details for. You can get device IDs from the search_hosts operation, the Falcon console, or the Streaming API. Maximum: 5000 IDs per request."
        ),
        view: Literal["summary", "full"] = Field(
            default="summary",
            description="'summary' returns investigation-essential fields only (default). 'full' returns the complete API response.",
        ),
        fields: list[str] | None = Field(
            default=None,
            description="Override: explicit list of fields to return. Takes precedence over view.",
        ),
        format: Literal["json", "toon"] = Field(
            default="json",
            description="Response format. 'toon' uses compact tabular encoding for token efficiency.",
        ),
    ) -> list[dict[str, Any]] | dict[str, Any] | str:
        """Retrieve detailed information for one or more host device IDs.

        Use when you already have specific device IDs from search results, the Falcon
        console, or the Streaming API. For discovering hosts by criteria, use
        falcon_search_hosts instead. Returns comprehensive host details.
        """
        # Resolve FieldInfo defaults for direct calls (e.g., from tests)
        if isinstance(view, FieldInfo):
            view = view.default
        if isinstance(fields, FieldInfo):
            fields = fields.default
        if isinstance(format, FieldInfo):
            format = format.default

        logger.debug("Getting host details for IDs: %s", ids)

        # Handle empty list case - return empty list without making API call
        if not ids:
            return []

        result = self._base_get_by_ids(
            operation="PostDeviceDetailsV2",
            ids=ids,
            id_key="ids",
        )

        if self._is_error(result):
            return result

        if isinstance(result, list):
            if fields:
                result = filter_records(result, fields)
            elif view == "summary":
                result = filter_records(result, HOST_SUMMARY_FIELDS)
            return format_response(result, format)

        return result

File: go/internal/toolsets/hosts/hosts.go
// Package hosts provides tools for searching and inspecting Falcon-managed
// hosts. It is the reference two-step search module: query device IDs honoring
// the requested sort, then hydrate full details by ID and restore that order.
package hosts

import (
	"context"
	_ "embed"
	"encoding/json"

	"github.com/crowdstrike/gofalcon/falcon/client"
	fhosts "github.com/crowdstrike/gofalcon/falcon/client/hosts"
	"github.com/crowdstrike/gofalcon/falcon/models"
	"github.com/google/jsonschema-go/jsonschema"

	fal "github.com/crowdstrike/falcon-mcp/internal/falcon"
	"github.com/crowdstrike/falcon-mcp/internal/toolsets"
)

//go:embed search_fql.md
var searchHostsFQLGuide string

// scopeHostsRead is the API scope both hosts operations require.
var scopeHostsRead = fal.Scope{Name: "Hosts", Read: true}

// defaultSearchLimit matches the Python tool's default of 10 records.
const defaultSearchLimit = 10

func init() { toolsets.Register("hosts", New) }

type searchHostsInput struct {
	Filter string `json:"filter,omitempty" jsonschema:"FQL filter expression. See falcon://hosts/search/fql-guide for syntax."`
	Limit  int64  `json:"limit,omitempty"  jsonschema:"The maximum records to return. [1-5000]"`
	Offset int64  `json:"offset,omitempty" jsonschema:"The offset to start retrieving records from."`
	Sort   string `json:"sort,omitempty"   jsonschema:"Sort by e.g. hostname.asc, last_seen.desc (also accepts hostname|desc)."`
}

// searchHostsInput satisfies toolsets.Constrainer (value receiver) so a drift
// in the interface surfaces as a build error rather than a silent no-op. A
// module using a pointer receiver would assert &fooInput{} instead.
var _ toolsets.Constrainer = searchHostsInput{}

// ApplyConstraints sets the limit bounds and default the Python tool declares
// (default=10, ge=1, le=5000), which jsonschema struct tags cannot express.
func (searchHostsInput) ApplyConstraints(schema *jsonschema.Schema) {
	lim := schema.Properties["limit"]
	if lim == nil {
		return
	}
	minLimit, maxLimit := 1.0, 5000.0
	lim.Minimum = &minLimit
	lim.Maximum = &maxLimit
	lim.Default = json.RawMessage("10")
	lim.Examples = []any{10, 100}
}

type getHostDetailsInput struct {
	IDs []string `json:"ids" jsonschema:"Host device IDs to retrieve details for. Maximum: 5000 IDs per request."`
}

// New builds the hosts toolset from an authenticated Falcon client.
func New(c *client.CrowdStrikeAPISpecification) *toolsets.Toolset {
	h := &handlers{c: c}
	return &toolsets.Toolset{
		Name:        "hosts",
		Description: "Search and inspect Falcon-managed hosts.",
		Tools: []toolsets.Tool{
			toolsets.NewTool("falcon_search_hosts", searchHostsDescription, toolsets.ReadOnly(), h.searchHosts),
			toolsets.NewTool("falcon_get_host_details", getHostDetailsDescription, toolsets.ReadOnly(), h.getHostDetails),
		},
		Resources: []toolsets.Resource{{
			URI:         "falcon://hosts/search/fql-guide",
			Name:        "falcon_search_hosts_fql_guide",
			Description: "Contains the guide for the `filter` param of the `falcon_search_hosts` tool.",
			MIMEType:    "text/markdown",
			Text:        searchHostsFQLGuide,
		}},
	}
}

const searchHostsDescription = "Search for hosts in your CrowdStrike environment.\n\n" +
	"Use this to find devices by hostname, platform, IP, sensor version, or other " +
	"attributes. Consult falcon://hosts/search/fql-guide before constructing filter " +
	"expressions. Returns full host details including device info, OS, and network context."

const getHostDetailsDescription = "Retrieve detailed information for one or more host device IDs.\n\n" +
	"Use when you already have specific device IDs from search results, the Falcon " +
	"console, or the Streaming API. For discovering hosts by criteria, use " +
	"falcon_search_hosts instead. Returns comprehensive host details."

type handlers struct {
	c *client.CrowdStrikeAPISpecification
}

// searchHosts queries device IDs by filter, then hydrates full details and
// restores the query-step sort order on the result.
func (h *handlers) searchHosts(ctx context.Context, in searchHostsInput) (any, error) {
	limit := in.Limit
	if limit == 0 {
		limit = defaultSearchLimit
	}

	q, err := h.c.Hosts.QueryDevicesByFilter(&fhosts.QueryDevicesByFilterParams{
		Context: ctx,
		Filter:  fal.Opt(in.Filter),
		Limit:   &limit,
		Offset:  fal.Opt(in.Offset),
		Sort:    fal.Opt(in.Sort),
	})
	if e := fal.APIError(err, q, scopeHostsRead); e != nil {
		return []any{e}, nil // parity: hosts wraps the search error in a list
	}

	ids := q.Payload.Resources
	if len(ids) == 0 {
		return []any{}, nil // parity: bare empty list
	}

	details, e := h.fetchDetails(ctx, ids)
	if e != nil {
		return []any{e}, nil
	}
	// Restore the query-step sort order in case the details endpoint returns
	// entities in a different order.
	return fal.ReorderByIDs(ids, details, deviceID), nil
}

// getHostDetails hydrates details for explicit device IDs. An empty ids list
// short-circuits with a bare empty list and makes no API call.
func (h *handlers) getHostDetails(ctx context.Context, in getHostDetailsInput) (any, error) {
	if len(in.IDs) == 0 {
		return []any{}, nil
	}
	details, e := h.fetchDetails(ctx, in.IDs)
	if e != nil {
		return e, nil // parity: get_host_details returns the bare error dict
	}
	return details, nil
}

// fetchDetails calls the details endpoint for the given IDs, returning the
// device entities or a normalized error.
func (h *handlers) fetchDetails(ctx context.Context, ids []string) ([]*models.DeviceapiDeviceSwagger, *fal.Error) {
	d, err := h.c.Hosts.PostDeviceDetailsV2(&fhosts.PostDeviceDetailsV2Params{
		Context: ctx,
		Body:    &models.MsaIdsRequest{Ids: ids},
	})
	if e := fal.APIError(err, d, scopeHostsRead); e != nil {
		return nil, e
	}
	return d.Payload.Resources, nil
}

// deviceID extracts the device_id from a host entity for order restoration.
func deviceID(d *models.DeviceapiDeviceSwagger) string {
	if d == nil || d.DeviceID == nil {
		return ""
	}
	return *d.DeviceID
}
